Add com.bestaticpy.bestatic

tatsatb opened the pull request:
<!-- ⚠️  The submission PR must be against the `new-pr` branch ⚠️  -->

### Please confirm your submission meets all the criteria

<!-- Please replace each `[ ]` with `[X]` when the step is complete -->

- [X] Please describe the application briefly.
  <!-- insert the description here -->
Bestatic is a static-site generator, written in Python and Jinja2 templating engine. It is minimal, yet fully-featured. And it can be installed graphically, without worrying about dependencies and package managers. Like every other static-site generator, it can generate a complete website by processing a bunch of markdown files. As a result, it eliminates the need for server-side programming or databases. The site Bestatic generates in _output folder can be served from any web server or host.
- [X] The domain used for the application ID is [controlled by the application developers][appid-domain] and the [application id guidelines][appid] are followed.
- [X] I have read the and followed all the [Submission and licence requirements][reqs].
- [X] I have [built][build] and tested the submission locally.
- [X] I am an author/developer/upstream contributor of the project. If not, I contacted upstream developers about this submission. **Link:**


<!-- ⚠️  Don't modify anything below this line ⚠️  -->

[appid-domain]: https://docs.flathub.org/docs/for-app-authors/requirements#control-over-domain
[appid]: https://docs.flathub.org/docs/for-app-authors/requirements#application-id
[reqs]: https://docs.flathub.org/docs/for-app-authors/requirements
[build]: https://docs.flathub.org/docs/for-app-authors/submission/#before-submission

tatsatb commented:
bot, build com.bestaticpy.bestatic

tatsatb commented:
bot, build com.bestaticpy.bestatic

bbhtt commented on com.bestaticpy.bestatic.metainfo.xml:
This should be in the upstream repo.

tatsatb replied:
Do I have to put this (and PNG) in https://github.com/tatsatb/bestatic repo?

bbhtt replied:
Yes

tatsatb replied:
Done (also added the PNG).

bbhtt replied:
Please use the one from upstream now and remove them here.

bbhtt replied:
You can add them as url sources and use the raw.githubusercontent.com links, use it from a commit not `main` / `master`.

tatsatb replied:
I just updated the json file by adding the URL sources from the commit. Does it look good?

bbhtt commented on com.bestaticpy.bestatic.png:
This should be in upstream repo

bbhtt commented on com.bestaticpy.bestatic.yml:
```diff
@@ -0,0 +1,16 @@
+modules:
+  - python3-Bestatic.json
+finish-args:
+  - --share=ipc
+  - --socket=fallback-x11
+  # Wayland access
+  - --socket=wayland
+  # Needs to talk to the network:
+  - --share=network
+  # Needs to save files locally
+  - --filesystem=home
```
Are all of these actually needed for a console application?

bbhtt replied:
Why do you need display access and home access?

tatsatb replied:
I believe I need the home access so that the user can create a website directory anywhere inside the home (as this program read and write files from the current directory). I tested with no filesystem access and it failed.  

I guess I do not really need display access. I added that so that when user clicks on localhost:8080 on terminal, a browser window opens. I thought without x11/wayland this may not work.

bbhtt replied:
Don't think that'd require display access here. That just opens the browser outside the sandbox.

tatsatb replied:
Updated the manifest.

bbhtt commented:
bot, build com.bestaticpy.bestatic

bbhtt commented on com.bestaticpy.bestatic.metainfo.xml:
Remove this

tatsatb replied:
Done.

bbhtt commented on com.bestaticpy.bestatic.png:
Remove this

tatsatb replied:
Done.

bbhtt commented on python3-Bestatic.json:
```diff
@@ -0,0 +1,266 @@
+    "sources": [
+        {
+            "type": "file",
+            "url": "https://raw.githubusercontent.com/tatsatb/bestatic/3ca8e2d249bdfde3fe8b1f7cc6f3eb6e4671626e/flatpak/com.bestaticpy.bestatic.metainfo.xml"
```
Add `sha256`

tatsatb replied:
Added both sha256.

bbhtt commented on python3-Bestatic.json:
```diff
@@ -0,0 +1,266 @@
+        },
+        {
+            "type": "file",
+            "url": "https://raw.githubusercontent.com/tatsatb/bestatic/3ca8e2d249bdfde3fe8b1f7cc6f3eb6e4671626e/flatpak/com.bestaticpy.bestatic.png"
```
Add sha256

bbhtt commented:
bot, build com.bestaticpy.bestatic

bbhtt commented:
You have to lowercase the checksums.

tatsatb commented:
I just updated in lowercase. That is interesting though, I thought sha256 is case insensitive.

tatsatb commented:
bot, build com.bestaticpy.bestatic

bbhtt commented:
```
flatpak run com.bestaticpy.bestatic

Traceback (most recent call last):
  File "/app/bin/bestatic", line 8, in <module>
    sys.exit(main())
             ^^^^^^
  File "/app/lib/python3.11/site-packages/bestatic/bestatic.py", line 55, in main
    with open("config.yaml", mode="rb") as ft:
         ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
FileNotFoundError: [Errno 2] No such file or directory: 'config.yaml'
```

bbhtt commented:
You have to add a stock icon tag in Metainfo for the icon to work. https://docs.flathub.org/docs/for-app-authors/metainfo-guidelines/#icons

bbhtt commented:
Is there an example I can try it on to see how it works?

tatsatb commented:
> Is there an example I can try it on to see how it works?

Sure thing. Just the bestatic command is expected to result the error that you quoted. It should be used when site is ready to build.

You have to have a `config.yaml' file and a `theme` directory inside the working directory (as an example, you can use the ones available in the upstream GitHub repo). You can also run `bestatic quickstart` to create a `config.yaml` file from scratch. You still have to download the `themes` folder from the upstream repo, unless you want to manually create a theme from the scratch and put it in the `themes` directory. Thank you.

tatsatb commented:
> You have to add a stock icon tag in Metainfo for the icon to work. https://docs.flathub.org/docs/for-app-authors/metainfo-guidelines/#icons

I honestly do not need an icon to show up as a desktop entry. I just wanted to add an icon so that it is shown on the Flathub. Do I need to add the <icon> tag in that case?

bbhtt commented:
>  I just wanted to add an icon so that it is shown on the Flathub. Do I need to add the tag in that case?

Yes

tatsatb commented:
> > I just wanted to add an icon so that it is shown on the Flathub. Do I need to add the tag in that case?
> 
> Yes

I have added the <icon> tag in the metainfo file and also updated the json file accordingly with correct file link and hash.

bbhtt commented:
bot, build com.bestaticpy.bestatic

hfiguiere commented:
> I just updated in lowercase. That is interesting though, I thought sha256 is case insensitive.

both `sha256sum` and pypi checksum are in lowercase, fyi.

hfiguiere commented on com.bestaticpy.bestatic.yml:
```diff
@@ -0,0 +1,12 @@
+command: bestatic
+modules:
+  - python3-Bestatic.json
+finish-args:
```
finish args is always found above modules

tatsatb replied:
Done.

hfiguiere commented on com.bestaticpy.bestatic.yml:
```diff
@@ -0,0 +1,12 @@
+sdk: org.freedesktop.Sdk
+command: bestatic
+modules:
+  - python3-Bestatic.json
```
the main module should be in this file not in a separate one. The automatically generated mainifest for the python dependencies however should be kept in a separate file for easy regeneration.

tatsatb replied:
I do not have any other module than the python3-Bestatic. We just had to edit a bit that what flatpak-pip-generator produced as it could not resolve all the PyPI dependencies correctly. Do you want me to move anything specific from json file to yaml file?

bbhtt replied:
You should move the metainfo file and icon installation steps + sources here. Those will get overwritten when regenerating with pip generator.

tatsatb replied:
Done.

bbhtt commented:
bot, build com.bestaticpy.bestatic
